File: frontend/src/app/services/proposal/proposal-service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { firstValueFrom } from 'rxjs';

import { MyProposalPage, Proposal } from '../../models/proposal.model';
import { CreateProposalRequest } from '../../dto/proposal/create-proposal-request';
import { environment } from '../../../environments/environment';
import { ProposalStatus } from '../../enums/proposal-status';

@Injectable({
  providedIn: 'root'
})
export class ProposalService {

  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/proposals`;

  async getByTicketId(ticketId: number): Promise<Proposal[]> {
    return firstValueFrom(
      this.http.get<Proposal[]>(
        `${this.baseUrl}/ticket/${ticketId}`
      )
    );
  }

  async getMyProposals(statuses: ProposalStatus[], page = 0, size = 10): Promise<MyProposalPage> {
    let params = new HttpParams();
    for (const status of statuses) {
      params = params.append('status', status);
    }
    params = params.set('page', page.toString());
    params = params.set('size', size.toString());

    return firstValueFrom(
      this.http.get<MyProposalPage>(`${this.baseUrl}/my`, {
        params
      })
    );
  }

  async create(dto: CreateProposalRequest): Promise<Proposal> {
    return firstValueFrom(
      this.http.post<Proposal>(this.baseUrl, dto)
    );
  }

  async reject(proposalId: number): Promise<Proposal> {
    return firstValueFrom(
      this.http.patch<Proposal>(
        `${this.baseUrl}/${proposalId}/reject`,
        {}
      )
    );
  }

  async accept(proposalId: number): Promise<Proposal> {
    return firstValueFrom(
      this.http.patch<Proposal>(
        `${this.baseUrl}/${proposalId}/accept`,
        {}
      )
    );
  }
}

File: backend/backend/src/main/java/br/com/trampo/backend/port/service/proposal/ProposalService.java
package br.com.trampo.backend.port.service.proposal;

import br.com.trampo.backend.domain.Users;
import br.com.trampo.backend.domain.enums.StatusProposal;
import br.com.trampo.backend.dto.proposal.CreateProposalDto;
import br.com.trampo.backend.dto.proposal.MyProposalDto;
import br.com.trampo.backend.dto.proposal.ProposalDto;
import br.com.trampo.backend.dto.common.PageDto;

import java.util.List;

public interface ProposalService {
    ProposalDto create(CreateProposalDto dto, Users user);

    List<ProposalDto> findByTicketId(int ticketId, Users user);

    PageDto<MyProposalDto> findMyProposals(Users user, List<StatusProposal> statuses, int page, int size);

    ProposalDto accept(int proposalId, Users user);

    ProposalDto reject(int proposalId, Users user);
}
